### PR TITLE
Add occ command to expire the trashbin

### PR DESCRIPTION
**Pull request by DeepDiver1975**
@PVince81 @butonic

### CODE REVIEW

**Comment by mmattel** [comment]
Great !
Beside expiring the trashbin for a/all users, I would additionally propose the ability to see what and how much data will be deleted. With the current process I just say do it.
Therefore I propose:
- --dryrun to see what would be deleted and how much data would be freed - without doing it
- --verbose to see online what is deleted
- in the results for the user how much actually has been deleted (number and quantity)
- a summary when finished deleting x files for y users freeing z MB

File: apps/files_trashbin/lib/Command/ExpireTrash.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+				'remove deleted files of the given user(s), if no user is given all deleted files will be removed'
+			);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+
+		$maxAge = $this->expiration->getMaxAgeAsTimestamp();
+		if (!$maxAge) {
+			$output->writeln("No expiry configured.");
+//			return;
```
Comment (PVince81) [inline apps/files_trashbin/lib/Command/ExpireTrash.php]: stray return ?

**Comment by PVince81** [comment]
@DeepDiver1975 what's there to discuss ? I heard the primary reason for adding this would be to be able to run OCC commands from cron to ease the load from cron ?

**Comment by PVince81** [comment]
CC @cdamken

File: apps/files_trashbin/lib/Command/ExpireTrash.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/files_trashbin/lib/Command/ExpireTrash.php]: :fearful:

**Comment by cdamken** [comment]
Awesome!! Thanks Guys!

@DeepDiver1975 I think this https://github.com/owncloud/enterprise/issues/1279 is would be fixed with this PR too

**Comment by cdamken** [comment]
@carlaschroder This should be added in the documentation ;)

**Comment by DeepDiver1975** [comment]
> @carlaschroder This should be added in the documentation ;)

fore sure - once we agree on this ...

**Comment by cdamken** [comment]
@DeepDiver1975 how about the versions?, could be added too?

**Comment by DeepDiver1975** [comment]
> how about the versions?, could be added too?

on it ....

**Comment by cdamken** [comment]
> on it ....

I own you some beers ;)

**Comment by DeepDiver1975** [comment]
@mmattel thanks for your input - we can enhance this over time in follow up prs.

**Comment by mmattel** [comment]
Ok to me,
pls consider that we use a different way defining users in eg files:scan (like I proposed above) than we do here. imho this should be consistant. would make future improvments also be consistant to other commands...

File: apps/files_trashbin/lib/Command/ExpireTrash.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+				InputArgument::OPTIONAL | InputArgument::IS_ARRAY,
+				'expires the trashbin of the given user(s), if no user is given the trash for all users will be expired'
+			);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+
+		$maxAge = $this->expiration->getMaxAgeAsTimestamp();
+		if (!$maxAge) {
+			$output->writeln("No expiry configured.");
```
Comment (PVince81) [inline apps/files_trashbin/lib/Command/ExpireTrash.php]: I'm getting this now. I thought we had defaults ? There was a thing about expiring trashbin items whenever the trashbin size reached 50% of the total available size

Comment (PVince81) [inline apps/files_trashbin/lib/Command/ExpireTrash.php]: Maybe it's new behavior, I can see that this command simply reflects the same behavior as the background job, so it's ok. Would be good to get some input from @VicDeo

Comment (DeepDiver1975) [inline apps/files_trashbin/lib/Command/ExpireTrash.php]: As far as I understood the code this whole expiry is purly based on time ... but I maybe I missed something in this mess .....

**Comment by PVince81** [comment]
I tried setting "trashbin_retention_obligation" to "auto" but it still says it's unconfigured. A quick debug in the "Expiration" code shows that it keeps the max age to "no obligation". So it looks like the background job wouldn't expire the trashbin based on size even though the doc says it would when set to "auto".
Anyway, this is not related to this PR.

**Comment by PVince81** [comment]
ok, looks like the size-based behavior is another single-shot expire job that gets scheduled only after a deletion.

**Comment by PVince81** [comment]
This is the other expire job for size:
```
| 14 | OC\Command\CommandJob                             | "O:33:\"OCA\\Files_Trashbin\\Command\\Expire\":1:{s:39:\"\u0000OCA\\Files_Trashbin\\Command\\Expire\u0000user\";s:36:\"80cd743a-fcb2-1035-90e4-53fe065c9f7b\";}" |          0 |   1471881096 |           0 |
| 15 | OC\Command\CommandJob                             | "O:33:\"OCA\\Files_Trashbin\\Command\\Expire\":1:{s:39:\"\u0000OCA\\Files_Trashbin\\Command\\Expire\u0000user\";s:36:\"80cde582-fcb2-1035-90e5-53fe065c9f7b\";}" |          0 |   1471881193 |           0 |
```

Looks like it would be too tricky to add this behavior into this command as well.

Ok ok, I'll now test with numeric values.

**Comment by PVince81** [comment]
Tested, works :+1: 

(it was really painful to manipulate timestamps in oc_filecache as I forgot to also re-set the path_hash since the timestamp is also in the filename)

**Comment by PVince81** [comment]
Ah, that was the trashbin stuff.

Now testing versions, let me put the steps now:

1. Setup OC with Mysql or MariaDB
1. Set "versions_retention_obligation" to "auto, 1" in config.php
1. Open text editor
1. Edit and save three times to create versions
1. Go to the data folder of the user in "data/$user/files_versions"
1. Pick one of the versions file, in my case I picked "test.txt.v1471883185"
1. Rename the file to "test.txt.v1371883185" (changed one digit)
1. `update oc_filecache set name=replace('v1471883185', 'v1371883185');`
1. (and that's the moment I started to really hate the filecache and its info redundancy)
1. `update oc_filecache set path=replace(path, 'v1471883185', 'v1371883185');`
1. `update oc_filecache set path_hash = md5(path) where path_hash <> md5(path);`
1. `update oc_filecache set mtime=1371883185 where mtime=1471883185;`
1. `select * from oc_filecache where name like '%v1371883185';` => see the entry exists
1. Run `occ versions:expire`
1. `select * from oc_filecache where name like '%v1371883185';` => entry is gone and file is gone from data dir

**Comment by PVince81** [comment]
@DeepDiver1975 fixed a bad copy-paste mistake to make version expire work: https://github.com/owncloud/core/pull/25878/commits/3d618a1cc5f6fe0605b98e1ade60481384daf664

**Comment by PVince81** [comment]
Both trashbin expire and version expire work now :+1:

File: apps/files_trashbin/lib/Command/ExpireTrash.php (hunk 2 above)
Comment (VicDeo) [inline apps/files_trashbin/lib/Command/ExpireTrash.php]: @PVince81 as per https://github.com/owncloud/core/issues/16300 that I documented in https://github.com/owncloud/core/blob/master/config/config.sample.php#L415 and 
`auto` guarantees neither keeping nor deletion. Items might be deleted if a disk usage is over 50% of quota. 
While exact values  specify number of days to store / to delete respectively.
From that point `auto` value means the mode before retention obligation was implemented. It is limited by quota usage for  trashbin and by by number of copies per different time ranges for versions

Comment (VicDeo) [inline apps/files_trashbin/lib/Command/ExpireTrash.php]: Anyway the best description was provided by @MTRichards in #16300 for Trashbin and #16301 for Versions. 
This is actually a reference spec for retention

**Comment by butonic** [comment]
can we make https://github.com/owncloud/core/blob/master/apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php#L34 reuse the new commend? same with versions? otherwise we just duplicate code.

**Comment by butonic** [comment]
👍 deduplicating would need further tweaking of the CommandJob. And increase complexity.
backport request to 9.0

**Comment by DeepDiver1975** [comment]
> deduplicating would need further tweaking of the CommandJob.

yes - there are many similarities in the way we implement occ commands, bachgroundjobs and repairsteps. We should think about implementing on single interface which can be reused in all three areas.

**Comment by DeepDiver1975** [comment]
preparing backport ...

**Comment by DeepDiver1975** [comment]
- stable9.1: https://github.com/owncloud/core/pull/25914

**Comment by cdamken** [comment]
Nice work! :+1:

**Pull request merged**